sycl/include/sycl/ext/intel/experimental/esimd/memory.hpp

```diff
@@ -1578,17 +1628,40 @@ __ESIMD_API void lsc_slm_block_store(uint32_t offset,
   constexpr uint16_t _AddressScale = 1;
   constexpr int _ImmOffset = 0;
   constexpr lsc_data_size _DS = detail::finalize_data_size<T, DS>();
-  static_assert(_DS == lsc_data_size::u32 || _DS == lsc_data_size::u64,
-                "Transposed store is supported only for data size u32 or u64");
-  constexpr detail::lsc_vector_size _VS = detail::to_lsc_vector_size<NElts>();
+
+  constexpr int SmallIntFactor =
+      (_DS == lsc_data_size::u16) ? 2 : (_DS == lsc_data_size::u8 ? 4 : 1);
+  static_assert(NElts > 0 && NElts % SmallIntFactor == 0,
+                "Number of elements is not supported by Transposed load");
+
+  constexpr int FactoredNElts = NElts / SmallIntFactor;
+  constexpr lsc_data_size ActualDS = sizeof(T) == 8
+                                         ? __ESIMD_ENS::lsc_data_size::u64
+                                         : __ESIMD_ENS::lsc_data_size::u32;
+
+  // Prepare template arguments for the call of intrinsic.
+  using StoreElemT = __ESIMD_DNS::__raw_t<
+      std::conditional_t<SmallIntFactor == 1, T, uint32_t>>;
+
+  constexpr detail::lsc_vector_size _VS =
+      detail::to_lsc_vector_size<FactoredNElts>();
   constexpr detail::lsc_data_order _Transposed =
       detail::lsc_data_order::transpose;
   constexpr int N = 1;
   __ESIMD_NS::simd_mask<N> pred = 1;
   __ESIMD_NS::simd<uint32_t, N> offsets = offset;
-  __esimd_lsc_store_slm<T, cache_hint::none, cache_hint::none, _AddressScale,
-                        _ImmOffset, _DS, _VS, _Transposed, N>(
-      pred.data(), offsets.data(), vals.data());
+  if constexpr (std::is_same_v<T, StoreElemT>) {
+    __esimd_lsc_store_slm<T, cache_hint::none, cache_hint::none, _AddressScale,
+                          _ImmOffset, _DS, _VS, _Transposed, N>(
```

Comment:
Here it may be better to use ActualDS instead of _DS. It doesn't change anything, but makes the code more uniform and eliminates doubts/thoughts about what is _DS in one of if-branches.